File: c++/1131.cpp
//
//  main.cpp
//  test
//
//  Created by 李寻欢 on 2019/10/16.
//  Copyright © 2019 李寻欢. All rights reserved.
//
#include <stdio.h>
#include <iostream>
#include <string>
#include <map>
#include <cmath>
#include <algorithm>
#include <stack>
#include <vector>
#include <queue>
#include <unordered_map>
#include <unordered_set>
#include <string.h>
#include <set>
#include <deque>
using namespace std;

int n,m,k;
int start,destination,next_station;
unordered_map<int, vector<int>> graph;
int g[10000][10000] = {0};
unordered_map<int, vector<int>> foot;
bool vis[10000] = {false};
vector<int> best_line;

vector<int> path;
vector<int> best_path;
vector<int> trans_node;
vector<int> best_trans;
void dfs(int s,int step,int &min_step,int trans,int &min_trans,int pre_line,vector<int> lines){
    if(vis[s] == true || step > min_step) return;
    if(s == destination){
        if(min_step > step){
            min_step = step;
            best_line = lines;
            min_trans = trans;
            best_path = path;
            best_trans = trans_node;
        }
        else if(min_step == step){
            if(trans < min_trans){
                min_trans = trans;
                best_line = lines;
                best_path = path;
                best_trans = trans_node;
            }
        }
        return;
    }
    vis[s] = true;
    path.emplace_back(s);
    for(auto &each:graph[s]){
        int back_up_line = pre_line;
        int cur_line = g[s][each];
        bool flag = false;
        if(pre_line != cur_line){
            flag = true;
            trans_node.emplace_back(s);
            lines.push_back(cur_line);
            pre_line = cur_line;
            trans++;
        }
        dfs(each, step+1,min_step,trans,min_trans,pre_line,lines);
        if(flag){
            lines.pop_back();
            trans--;
            pre_line = back_up_line;
            trans_node.pop_back();
        }
    }
    vis[s] = false;
    path.pop_back();
    
}

int main(){
    scanf("%d",&n);
    for(int i=1;i<=n;++i){
        scanf("%d%d",&m,&start);
        for(int j=1;j<m;++j){
            scanf("%d",&next_station);
            g[start][next_station] = g[next_station][start] = i;
            graph[start].emplace_back(next_station);
            graph[next_station].emplace_back(start);
            start = next_station;
        }
    }
    scanf("%d",&k);
    for(int i=0;i<k;++i){
        foot.clear();
        best_line.clear();
        fill(vis, vis+10000, false);
        scanf("%d%d",&start,&destination);
        int min_step = 10000;
        int min_trans = 10000;
        vector<int> lines;
        dfs(start, 0, min_step, 0, min_trans, -1, lines);
        printf("%d\n",min_step);
        for(int i=0;i<min_trans;++i){
            printf("Take Line#%d from %04d to ",best_line[i],best_trans[i]);
            if(i != min_trans-1)
                printf("%04d.\n",best_trans[i+1]);
            else
                printf("%04d.\n",destination);
        }
        
    }
}

File: Python3/1131.py
n=int(input())
lines = [None]
which_line = {}
graph = {}
vis = {}
for i in range(1,n+1):
    _,*line = map(int,input().split())
    a = line[0]
    vis[a] = False
    for each in line[1:]:
        vis[each] = False
        which_line[(a,each)] = i
        which_line[(each,a)] = i
        if graph.get(a) is None:
            graph[a] = []
        if graph.get(each) is None:
            graph[each] = []
        graph[a].append(each)
        graph[each].append(a)
        a = each
    lines.append(line)

def dfs(start,destination,step,pre_line):
    global vis,temp_path,min_step,path,trans_line,best_line,trans_node,node
    if(vis[start] or step > min_step): return
    if start == destination:
        if step < min_step:
            min_step = step
            path = temp_path[:]
            best_line = trans_line[:]
            node = trans_node[:]
        elif step == min_step and len(trans_line) < len(best_line):
            best_line = trans_line[:]
            path = temp_path[:]
            node = trans_node[:]
        return
    vis[start] = True
    temp_path.append(start)
    for next_station in graph[start]:
        origin_line = pre_line
        cur = which_line[(start,next_station)]
        flag = False
        if cur != pre_line:
            trans_line.append(cur)
            pre_line = cur
            flag = True
            trans_node.append(start)
        dfs(next_station,destination,step+1,pre_line)
        if flag:
            pre_line = origin_line
            trans_line.pop()
            trans_node.pop()
    temp_path.pop()
    vis[start] = False

k = int(input())
for _ in range(k):
    for each in vis:
        vis[each] = False
    start,destination = map(int,input().split())
    min_step = 10000
    temp_path = []
    path = []
    trans_line = []
    best_line = []
    node = []
    trans_node=[]
    dfs(start,destination,0,-1)
    print(min_step)
    for i in range(len(node)):
        print("Take Line#{} from {:04d} to ".format(best_line[i],node[i]),end='')
        if i != len(node) - 1:
            print("{:04d}.".format(node[i+1]))
        else:
            print("{:04d}.".format(destination))
